Just for comments (#1)
potatoprank/LuauBoops: justforPRreviewpurposes.md

# Introduction 
[Luau](https://luau.org/) is an open source general purpose programming language that is developed and maintained by [Roblox](https://roblox.com). Luau is a superset of [Lua](https://www.lua.org/) and introduce many new features most notably typing! It is often used to develop experiences for [Roblox](https://roblox.com). This is guide is mainly written for folks who have some familiarity with programming and would like to start using Luau! If you are new to programming and would like to learn Luau for the purpose of writing your own Roblox game, we suggest you get started on the [creator docs site](https://create.roblox.com/docs)!

> We aim to keep as much of this content as general as accessible as possible and many parts will explicity mention how to use Luau with Roblox. Sections that are entirely new to Luau are tagged with `Luau`. If you are skimming through the guide because you are already familiar with Lua, these are the sections you want to pay attention to!

Reviewer: ```suggestion
We aim to keep as much of this content as general and accessible as possible and many parts will explicitly mention how to use Luau with Roblox. Sections that are entirely new to Luau are tagged with `Luau`. If you are skimming through the guide because you are already familiar with Lua, these are the sections you want to pay attention to!
```

Reviewer: typos

This guide was started as a Roblox Hack Week project. We aim to grow and evolve this guide over time and the current version should already be more than enough to really get you going with Luau! Please open issues or PRs on github <TODO link> if you have any ideas to improve the guide!

# Additional Resources
- [Lua 5.1 manual](https://www.lua.org/manual/5.1/manual.html#5.4) (5.1 is the version where Luau diverged from Lua).
- [Luau lang website](https://luau.org/library)
- [Roblox Creator Docs site](https://create.roblox.com/docs)

# Getting Started
While Luau can be used as a standalone language, it is almost always embedded as a scripting language in another program such as Roblox. The easiest way to get started with writing and running Luau code is with the [Luau online interpreter](https://luau.org/demo). The next easiest way is...

## With Roblox

> To get started with Luau in Roblox, go to the [Roblox Creator Hub](https://create.roblox.com/), log in/make an account, and download Roblox Studio. From the templates tab, create a new empty Baseplate project. Then in the explorer widget, write click under "workspace" and create a new script. To test your script, press the ▶️ button and you should see any output (from `print` commands which we will introduce later) in the Console widget.

Reviewer: ```suggestion
To get started with Luau in Roblox, go to the [Roblox Creator Hub](https://create.roblox.com/), log in/make an account, and download Roblox Studio. From the templates tab, create a new empty Baseplate project. Then in the explorer widget, right-click under "workspace" and create a new script. To test your script, press the ▶️ button and you should see any output (from `print` commands which we will introduce later) in the Console widget.
```

Reviewer: more typos

The built in Roblox script editor will automatically run the Luau typechecker on your code and tell you if there are any errors

<TODO screenshot>

> If you are using tools like LFS or Rojo to develop code in Visual Studio Code, you can use the community created [Luau Language Server](https://marketplace.visualstudio.com/items?itemName=JohnnyMorganz.luau-lsp) extension to get the same typechecking experience in your editor!

Reviewer: > tools like LFS or Rojo

Suggest linking the homepages to these tools

Reviewer: For Rojo, you can link it to the section you made below

Author: done thanks!

Luau scripts are wrapped inside of `Instances` in the Roblox engine.

There are 3 types of script `Instances`

- Scripts: these run on the server only
- LocalScripts: these run on clients only
- ModuleScripts: these do not run by themselves

Anytime a `Script` or `LocalScript` instance is created (for example, when you launch the experience), the Roblox engine will run the code in the file. You can not control the order in which scripts run. These are the entry points into your Luau program.

`ModuleScripts` are reusable modules can be `require(...)`'d by other scripts. One major difference when developing on Roblox is that the Roblox runtime environment has a Datamodel rather than a filesystem. Therefore, instead of traversing the filesystem e.g. `require(helpers/myhelperscript.luau)` you need to traverse the Datamodel to find your module script e.g. `require(GetService("ReplicatedStorage").helpers.myhelperscript)`

TLDR to run Luau on Roblox, create a `Script` instance and hit the playtest button. Your script will run on the server and you will still be able to see its output on your local console.

# Getting Started

# Basic 💩

## Variables

> Variables are about what you'd expect in Luau. Variables can be assigned to with the `=` operator. 

Reviewer: ```suggestion
Variables are about what you'd expect in Luau. Variables are assigned with the `=` operator. 
```

> Global variables do not need to be declared. You simple assign to them and they get created if they don't already exist or assigned to if they do.

Reviewer: ```suggestion
Global variables do not need to be declared. You simply assign them, and they're created if they don't already exist or assigned if they do.
```

```
myGlobalVar = "I created a global variable"
```

To remove a global variable, you simply assign `nil` to it

```
myGlobalVar = nil
```

> In general, an object with a `nil` value  is equivalent to object not existing in Luau. You will see more of this in the tables section.

Reviewer: ```suggestion
In general, an object with a `nil` value  is equivalent to an object not existing in Luau. You will see more of this in the tables section.
```

### Enabling the type checker

In Roblox, the Luau type checker is set to run in nonstrict mode by default. We recommend enabling strict mode by adding the `--!strict` directive to the top of your luau script. If you are using Luau outside of Roblox or with a third party tool like Rojo, there may be other configuration files you can use to globally set what mode of the typechecker

<details>
    <summary>Luau Typechecker Modes (not important)</summary>

    - --!nocheck,
        - do not run the typechecker
    - --!strict
        - (recommended) run the type checker in strict mode
    - --!nonstrict
        - (default) similar to strict mode, but is more relaxed in inferring types.
</details>

### Basic syntax

To declare the type of a variable, use the `:` symbol. For example, to declare a variable of type `number` (types are covered later) we can do the following:

`local myTypedNumber : number`

> Variables in functions are type in the same way, and the return type is added at the end:

Reviewer: ```suggestion
Variables in functions are typed in the same way, and the return type is added at the end:
```

```
local function parrot(wordToSay : string, numberTimes : string) : string
    local output : string = ""
    for i = 1, numberTimes, 1 do
        output = output .. wordToSay
    end
    return output
end

```

Luau will also infer types for you.

```
local x = 5
print(typeof(x)) -- prints "number"
```

It is this author's opinion that you should avoid relying on type inference as much as possible!

> <sub>NOTE There are some limitations in the current Luau type checker when it comes to type inference and type refinement. There is a new typechecker (to be released someday) that will solve many of these issues!</sub>

Reviewer: A big limitation right now is type refinement for union-ed table types, see https://roblox.slack.com/archives/C6JT37EF8/p1729624732842919

print(b) -- prints "world"
```

Note that all value types are also immutable. The only other immutable references in Luau are frozen tables. These are tables that have been made immutable with the `table.freeze` function. We will talk about them more in the table section. 

# Tables (Introduction)
Tables are the do-all data structure in Lua. While there are a few more options in Luau, tables remain super important so it's important to understand them well! 

## Basics
Tables are associative arrays (AKA dictionaries/maps) and can be indexed with strings, numbers, or any other Luau value except `nil`!

To create an empty table:

```
local myTable = {}
```

To set or modify values in a table:

```
myTable.name = "Alice"
myTable.age = 25
``` 

To create a table with some initial values:

```
local myTable = {name = "Alice", age = 25}
```

To access a value in a table:

```
print(myTable.name) -- prints "Alice"
print(myTable.age) -- prints 25
```

To explicitly define the type of the above table:

```
local myTable : {name: string, age: number} = {name = "Alice", age = 25}
```

To delete a key from a table:

```
myTable.name = nil
```

To check if a key exists in a table:

```

> if myTable.name != nil then

Reviewer: ```suggestion
if myTable.name ~= nil then
```

for i, value in ipairs(myTable) do
    print(i, value)
end
-- prints "1 a" followed by "2 b" followed by "3 c"
```

## Table Keys
In the above examples, we used built in Luau syntax that automatically defines literal numeric and string keys. However, you can use any Luau value as a key in a table including values contained in variables. The generic syntax for accessing a value in a table is `myTable[key]` where `key` is some luau value.

```
local myTable = {}
local key = "name"
myTable[key] = "Alice"
print(myTable[key]) -- prints "Alice"
print(myTable.name) -- prints "Alice"
myTable["age"] = 25
print(myTable["age"]) -- prints 25
print(myTable.age) -- prints 25
```

You can even use other tables as table keys

```
local someTable = {}
local someOtherTable = {}
local myTable = {}
myTable[someTable] = "hello"
myTable[someOtherTable] = "world"
print(myTable[someTable]) -- prints "hello"
print(myTable[someOtherTable]) -- prints "world"
``` 

You can also use this syntax when constructing tables
    
```
local nameKey = "name"
local ageKey = "age"
local myTable = {
    [nameKey] = "Alice",
    [ageKey] = 25
}
print(myTable.name, myTable.age) -- prints "Alice 25"
```

## As Objects

> Tables are often used as objects in Lua. A common pattern is to create an explicitly typed table to representy our object and create a set of functions that operate on that table. 

Reviewer: ```suggestion
Tables are often used as objects in Lua. A common pattern is to create an explicitly typed table to represent your object and create a set of functions that operate on that table. 
```

```
Instance myInstance = workspace.FindFirstChild("bearModel")
myInstance.Name = "bear" -- `Name` is a property of the `Instance` datatype so this is valid
myInstance.location = "Canada" -- `location` is not a property of myInstance, this is not valid!
```

Roblox data types are all documented on the very useful [Roblox Creator Docs](https://create.roblox.com/docs/reference/engine) site.

## vector (new to Luau)

The Luau vector type represents a 3D vector. This is in fact the same type as the differently named [`Vector3` type in Roblox](https://create.roblox.com/docs/reference/engine/datatypes/Vector3) although currently the type checker thinks they are different (we'll fix it soon). You can use either the bulit in vector library or the Roblox API variants. The vector library variants are more portable!

> The vector type is either 3 or 4 dimensions (x,y,z + w) depending on what the compile time `LUA_VECTOR_SIZE` flag is set to. 3 component vectors work like you would expect representing x y and z components. Roblox is configered to use 3 component vectors. 4 component vectors are needed for affine transformations which allow for things like translations to be represented as a matrices. This is super duper useful for computer graphics and much less useful for most games on Roblox. 4 component vectors are not useable on Roblox and thus are not covered in this guide.

Reviewer: ```suggestion
The vector type is either 3 or 4 dimensions (x,y,z + w) depending on what the compile time `LUA_VECTOR_SIZE` flag is set to. 3 component vectors work like you would expect representing x y and z components. Roblox is configured to use 3 component vectors. 4 component vectors are needed for affine transformations which allow for things like translations to be represented as a matrices. This is super duper useful for computer graphics and much less useful for most games on Roblox. 4 component vectors are not useable on Roblox and thus are not covered in this guide.
```

To learn more about the vector library, see the [Luau Vector Library](https://luau.org/library#vector-library) page.

## buffer (new to Luau)

The buffer is a fixed size mutable (as in you can modify it's contents) block of memory. All buffer operations are done through the built in `buffer` library. Buffers are useful in applications where bandwidth or performance is critical. Otherwise consider using tables as arrays instead. Note that all buffer library operations convert the contents to and from the `number` or `string` type but the internal representation is still a byte array.

To learn more about the buffer library, see the [Luau Buffer Library](https://luau.org/library#buffer-library) page.

## thread 

> The `thread` type are used to represent a coroutine. These can be manipulated with the builtin [coroutine library](https://www.lua.org/pil/9.html). The Programming in Lua book has a [good section](https://www.lua.org/pil/9.html) on coroutines if you are interested in learning more about them.

Reviewer: ```suggestion
The `thread` type represents a coroutine. These can be manipulated with the builtin [coroutine library](https://www.lua.org/pil/9.html). The Programming in Lua book has a [good section](https://www.lua.org/pil/9.html) on coroutines if you are interested in learning more about them.
```

If you are working in Roblox, coroutines are often managed by the [task library](https://create.roblox.com/docs/reference/engine/libraries/task). These methods create coroutines that are managed by the Roblox engine's TaskSchedule such that should not (ever) manually `coroutine.yield` and `coroutine.resume` them. The `thread` type and the `coroutine` type mentioned on the Roblox API website are the same.

# More Luau Features
The [Luau website](https://luau.org/getting-started) documents most of the important new features of Luau. Unfortunately, this site doesn't outline all the features and there is no comprehensive reference manual for Luau yet (there really ought to be). 

Luau is an actively developed language and new features are being added all the time. You can propose and discuss new features on the [Luau RFCs github page](https://github.com/luau-lang/rfcs).

# Luau Tooling
Luau is open source and has a number super useful community supported tooling! This section is rather opinionated.

> ## Rojo

Reviewer: Link to Rojo homepage here like you did for Wally

## Wally 

[Wally](https://wally.run/) is a Luau package manager similar to NPM. Used together with Rojo, this is perhaps the best way to get external Luau dependencies into your Roblox project.

Note some Roblox developers use NPM to manage their Luau dependencies. I can't comment on this workflow as I don't use it myself. Wally seems good enough for me.

## Aftman/Foreman/Rokit

For *reasons* there are 3 nearly identical toolchain managers for Luau. [Rokit](https://github.com/rojo-rbx/rokit) is probably the best for being the most compatible but all 3 will work fine. These tools are useful for managing your Rojo and Wally installations.

## Luau Language Server

> [The Luau Language Server extension](https://github.com/JohnnyMorganz/luau-lsp) for WSCode provides syntax highlighting, code completion, and type checking for Luau code. This is a must have for any Luau developer! Note there are a couple similar extensions. The above one seems to be the best at this moment.

Reviewer: ```suggestion
[The Luau Language Server extension](https://github.com/JohnnyMorganz/luau-lsp) for VSCode provides syntax highlighting, code completion, and type checking for Luau code. This is a must have for any Luau developer! Note there are a couple similar extensions. The above one seems to be the best at this moment.
```

## Lune 

[Lune](https://github.com/lune-org/lune) is a standalone Luau runtime. It is the Luau analogy to Node.js. This is a great tool for running Luau code outside of Roblox. Some developers use this for writing tests that aren't tied to the Roblox engine. Note that Roblox Luau does not support path requires so you may need to use something like [darklua](https://darklua.com/) to transform your lune code to work with Roblox.

Please see the [Third Party Tools](https://create.roblox.com/docs/projects/external-tools) for using Luau tooling with Roblox.

# Roblox UI in Luau
## native
All of Roblox's [UI instances](https://create.roblox.com/docs/ui) can be managed with Luau code allowing you to generate UI entirely from Luau. This is the simplest create UI in Roblox. Many creators will author their UI in the Roblox Studio editor and then add Luau code to manage the interactions. You can also create your UI entirely from Luau code.

> Using native UI is the simplest way to create UI in Roblox. You may find that you will start running into limitations as your UI needs become mor demanding and complex. Thankfully, there are UI libraries that can help you with this!

Reviewer: ```suggestion
Using native UI is the simplest way to create UI in Roblox. You may find that you will start running into limitations as your UI needs become more demanding and complex. Thankfully, there are UI libraries that can help you with this!
```
